enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-04
+last-updated: 2020-06-05
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
```

Comment:
This needs a whole subsection, with some possibilities picked up for this enhancement and some punted into alternatives and possibly picked up in future expansions.  Internally, we've discussed all of:

1. Run these right after the update is selected, as we currently do for preconditions since openshift/cluster-version-operator#243.
2. Run these on-demand for cluster admins, so they can do things like "I'm planning on updating to $VERSION in a month; give me a fresh preflight report to turn up any current issues".
3. Have the CVO proactively poll with some (configurable?) staleness threshold, like [this][1].

Those are not mutually exlcusive.  If folks buy [my argument][1] for reporting preflight results via a new custom resource, I'd make that custom resource itself a request for an audit.  Something like:

```yaml
apiVersion: config.openshift.io/v1alpha1
kind: UpdatePreflight
metadata:
  name: whatever-the-creator-wants
  namespace: openshift-cluster-version
spec:
  image: quay.io/openshift-release-dev/ocp-release@sha256:7613d8f7db639147b91b16b54b24cfa351c3cbde6aa7b7bf1b9c80c260efad06
  frequency: 1d # optional *Duration, https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1?tab=doc#Duration
status:
  ... report fields go here...
```

Then the `UpdatePreflight` controller could just process those, and admins could create them directly, or the CVO could create them, and the CVO could block on a pass for the requested target version before going ahead with the update.

[1]: https://github.com/openshift/enhancements/pull/363/files#r436953681

Reply:
@crawford Do you have any thoughts on how far we want to go here?

Reply:
Added a section https://github.com/openshift/enhancements/commit/4cadad51eb7b2e999146ae9e8af76e57f9c90556#diff-f53e26d63b28b30cadb2bd5bc6342812R80